Fixed #31516 -- Improved naming of migrations with multiple operations.

manav014 opened the pull request:
[ticket-31516](https://code.djangoproject.com/ticket/31516)

Supersedes [#12799  ](https://github.com/django/django/pull/12799)

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,24 @@ def suggest_name(self):
+            name = new_name
+            left_out -= 1
+        if(left_out > 0):
+            name = name + f"_+{left_out}"
```
We shouldn't use `+` in filenames, moreover the number of additional operations doesn't have much value.

`_and_more` sounds like a good marker (see [comment](https://github.com/django/django/pull/12799#issuecomment-636694424)).

felixxm replied:
> What about `_16_more`(16 is the number of operations)?

I don't see much value in this format,  you still need to open the file to find out more.

felixxm reviewed (commented):
@manav014 Thanks for this patch :+1: It looks like you missed my [comments](https://github.com/django/django/pull/12799#issuecomment-747898443).

felixxm commented:
@manav014 My [comments](https://github.com/django/django/pull/12799#issuecomment-747898443) are not solved. Please check them before marking the patch as ready for review.

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,24 @@ def suggest_name(self):
+        fragments = [name for name in raw_fragments if name]
+
+        if not fragments:
+            return f"auto_{get_migration_name_timestamp()}"
```
f-strings shouldn't contain function calls. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

khu5h1 commented:
@felixxm Thank you for your suggestions.

> IMO we should use "auto" names if any operation doesn't have a suggested name, because it can be confusing to return the same name for different set of operations
I have added a check for the operations with no suggested name.

> we could add some mark to inform that a migration contains other operations but their names were truncated (ellipsis is of course a bad idea):
I have added `_and_more` in the name of the migration file for this.

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
+        if not fragments or None in fragments:
+            return f"auto_{migration_name_timestamp}"
+        name = fragments[0]
+        left_out = len(fragments) - 1
+        for fragment in fragments[1:]:
+            new_name = name + f"_{fragment}"
+            if len(new_name) > 52:
+                break
+            name = new_name
+            left_out -= 1
+        if(left_out > 0):
+            name = name + "_and_more"
```
`left_out` is not necessary anymore:
```suggestion
        name = fragments[0]
        for fragment in fragments[1:]:
            new_name = f'{name}_{fragment}'
            if len(new_name) > 52:
                name = f'{name}_and_more'
                break
            name = new_name
```

manav014 replied:
I used it as a way to check for the need of `_and_more`. So in order to check for that I feel this is required. Is there any other optimized way to check for that which I am missing?

felixxm replied:
My comment contains a suggestion.

felixxm commented on tests/migrations/test_autodetector.py:
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
+    def test_operation_with_no_suggested_name(self):
+        class Migration(migrations.Migration):
+            operations = [
+                migrations.CreateModel("Person1", fields=[]),
```
Use single-quotes in the new code.

felixxm commented on tests/migrations/test_autodetector.py:
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
+
+    def test_single_create_model_long_name(self):
+        """
+        Check that the first migration_name_fragment values isn't cut when
```
To add more explanation, docstrings should state the expected behavior and omit prefixes like "Check that" since all tests check things. This guideline is from [Python coding style](https://docs.djangoproject.com/en/dev/internals/contributing/writing-code/coding-style/#python-style).

manav014 commented:
@felixxm Thank you for your suggestions.

>IMO we should use "auto" names if any operation doesn't have a suggested name, because it can be confusing to return the same name for different set of operations

I have added a check for the operations with no suggested name.

>we could add some mark to inform that a migration contains other operations but their names were truncated (ellipsis is of course a bad idea):

I have added `_and_more` in the name of the migration file for this.

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
+        raw_fragments = (op.migration_name_fragment for op in self.operations)
+        fragments = [name for name in raw_fragments]
+
+        migration_name_timestamp = get_migration_name_timestamp()
+        if not fragments or None in fragments:
+            return f"auto_{migration_name_timestamp}"
```
We can move `get_migration_name_timestamp()` call to `if`.

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,25 @@ def suggest_name(self):
+        fragments = [name for name in raw_fragments]
+
+        migration_name_timestamp = get_migration_name_timestamp()
+        if not fragments or None in fragments:
```
We should probably check for empty strings as well.

felixxm reviewed (commented):
@manav014 Thanks for updates :+1: I left more comments.

@adamchainz Do you want to take a look?

felixxm commented on django/db/migrations/migration.py:
```diff
@@ -184,15 +183,22 @@ def suggest_name(self):
-            name = '_'.join(sorted(o.migration_name_fragment for o in self.operations))
-        if name is None:
-            name = 'initial' if self.initial else 'auto_%s' % get_migration_name_timestamp()
+        if self.initial:
+            return 'initial'
```
I've moved the initial migrations changes to a separate PR, see #14116.

manav014 replied:
Acknowledged.
May you help me with the changes that I shall commit in [this PR](https://github.com/django/django/pull/14109)? Or Shall I just remove Conflicts(the `if condition` for the initial name)?

felixxm commented on tests/migrations/test_autodetector.py:
```diff
@@ -2663,16 +2663,83 @@ class Migration(migrations.Migration):
         suggest_name = migration.suggest_name()
         self.assertIs(suggest_name.startswith('auto_'), True)
 
+    def test_no_fragments(self):
+        class Migration(migrations.Migration):
+            operations = [migrations.RunSQL("SELECT 1")]
+        migration = Migration('some_migration', 'test_app')
+        self.assertTrue(migration.suggest_name().startswith('auto_'))
```
This is redundant with `test_none_name()`.

felixxm reviewed (commented):
@manav014 Thanks for updates :+1: 

I rebased from `main`, pushed edits, and moved tests for migrations with no operations to a separate commit.

adamchainz commented:
Thanks for picking this up and merging @manav014 @felixxm !
